lib/MetricsProducer.js

```diff
@@ -0,0 +1,68 @@
+'use strict'; // eslint-disable-line strict
+
+const async = require('async');
+const { Logger } = require('werelogs');
+
+const BackbeatProducer = require('./BackbeatProducer');
+const MetricsModel = require('./models/MetricsModel');
+
+class MetricsProducer {
+    /**
+    * @constructor
+    * @param {Object} zkConfig - zookeeper configurations
+    * @param {Object} mConfig - metrics configurations
+    */
+    constructor(zkConfig, mConfig) {
+        this._zkConfig = zkConfig;
+        this._topic = mConfig.topic;
+
+        this._producer = null;
+        this._log = new Logger('MetricsProducer');
+    }
+
+    setupProducer(done) {
+        const producer = new BackbeatProducer({
+            zookeeper: { connectionString: this._zkConfig.connectionString },
+            topic: this._topic,
+        });
+        producer.once('error', done);
+        producer.once('ready', () => {
+            producer.removeAllListeners('error');
+            producer.on('error', err => {
+                this._log.error('error from backbeat producer',
+                               { error: err });
+            });
+            this._producer = producer;
+            done();
+        });
+    }
+
+    /**
+     * @param {Object} extMetrics - an object where keys are all buckets for a
+     *   given extension and values are the metrics for the bucket
+     *   (i.e. { my-bucket: { ops: 1, bytes: 124 } } )
+     * @param {String} type - type of metric (queueud or processed)
+     * @param {String} ext - extension (i.e. 'crr')
+     * @return {undefined}
+     */
+    publishMetrics(extMetrics, type, ext) {
```

Comment:
I would advise to add a `cb` callback argument when all metrics have been effectively published. It can be optional, if the caller wants fire&forget, but in many cases it can be useful to know this. Otherwise, if you prefer to leave as is, make sure you add a comment in the JSDoc stating that the actual processing is asynchronous, so the caller does not expect the metrics to be effectively published as soon as the function returns.

Reply:
I've added a cb to this, but the changes only currently log an error message. Especially for `LogReader`, I didn't want error from publishing metrics to stop the current batch from completing processing. Maybe it's not something I should be worried about? Not exactly sure

Reply:
Even logging at the end is a good reason to provide a callback to the function. I agree we should not impact the main batch if the metrics cannot be published. The main point here is that allowing a user to be notified when the operation completes does no harm an allows a few things to happen (e.g. debug logging at the end, in case we need to troubleshoot), even though the critical path does not need to wait on this callback.

It may also be used to avoid stacking operations when there is already one in progress (if publish metrics for some reason takes more time than the interval between two batches). In such case it can be cleaner to defer the next publish operation. Not that we have to do this now, just mentioning as a possible use case for a callback.